src/kvp/mod.rs

```diff
@@ -0,0 +1,546 @@
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    fmt::Display,
+    marker::PhantomData,
+    ops::Deref,
+    str::FromStr,
+};
+
+use serde::{de::Visitor, ser::SerializeMap, Deserialize, Serialize};
+use snafu::{ensure, ResultExt, Snafu};
+
+mod annotation;
+pub mod consts;
+mod key;
+mod label;
+mod value;
+
+pub use annotation::*;
+pub use key::*;
+pub use label::*;
+pub use value::*;
+
+#[derive(Debug, PartialEq, Snafu)]
+pub enum KeyValuePairError<E>
+where
+    E: std::error::Error + 'static,
+{
+    #[snafu(display("label input cannot be empty"))]
+    EmptyInput,
+
+    #[snafu(display("invalid number of equal signs - expected exactly 1, got {signs}"))]
+    InvalidEqualSignCount { signs: usize },
+
+    #[snafu(display("failed to parse label key"))]
+    InvalidKey { source: KeyError },
+
+    #[snafu(display("failed to parse label value"))]
+    InvalidValue { source: E },
+}
+
+/// A [`KeyValuePair`] is a pair values which consist of a [`Key`] and value.
+/// These pairs can be used as Kubernetes labels or annotations. A pair can be
+/// parsed from a string with the following format: `(<PREFIX>/)<NAME>=<VALUE>`.
+///
+/// ### Links
+///
+/// - <https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/>
+/// - <https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/>
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct KeyValuePair<V>
```

Comment:
We can change the `KeyValuePair` struct to a newtype struct like: `KeyValuePair<V>((Key, V))`. I like having a dedicated type with which we can provide a developer friendly API.

Reply:
> I like having a dedicated type with which we can provide a developer friendly API.

What special APIs are you thinking about here?

`(K, V)` is a common Rust convention, and allows for seamless roundtripping with `HashMap`/`BTreeMap` (see [`Extend`](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#impl-Extend%3C(K,+V)%3E-for-HashMap%3CK,+V,+S%3E), [`From`](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#impl-From%3C%5B(K,+V);+N%5D%3E-for-HashMap%3CK,+V,+RandomState%3E), [`FromIterator`](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#impl-FromIterator%3C(K,+V)%3E-for-HashMap%3CK,+V,+S%3E), [`IntoIterator`](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#impl-IntoIterator-for-HashMap%3CK,+V,+S%3E), [`iter()`](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#method.iter), etc. I'm not saying that it can't be done, but I'd like to understand what you expect to gain from doing it.

Reply:
> What special APIs are you thinking about here?

For example `KeyValuePair::new()`, `kvp.key()` or `kvp.value()`. The custom type also "ecapsulates" specific behaviour and different impls like `Display` or `Serialize`.

To get the PR forward, an intermediate solution could be to implement `Deref<Target=(Key, V)> where V: Value` for `KeyValuePair`. We can then explore using a plain `(K, V)` in a future PR.

Reply:
> For example KeyValuePair::new(), kvp.key() or kvp.value().

So.. `(k, v)`, `kvp.0`, and `kvp.1`, respectively.

> The custom type also "ecapsulates" specific behaviour and different impls like Display or Serialize.

`Display` I can sort of understand, but what on earth do you need a non-standard `Serialize` for here? That just looks like a very special-cased workaround for abusing `BTreeSet` instead of `BTreeMap`.

Reply:
> So.. `(k, v)`, `kvp.0`, and `kvp.1`, respectively.

Yes, but the decision to use a dedicated type was inspired by this pattern: https://www.shuttle.rs/blog/2022/07/28/patterns-with-rust-types#contract-based-programming-in-rust--sub-typing

> `Display` I can sort of understand, but what on earth do you need a non-standard `Serialize` for here? That just looks like a very special-cased workaround for abusing `BTreeSet` instead of `BTreeMap`.

I don't know what `BTreeSet` and `BTreeMap` has to do with this? We are talking about a single `KeyValuePair`, which doesn't use any of these collection types. The serialization of the type works very similar to a map, because the identifier is part of the inner data instead the item (struct/enum) name serde uses per default.

Reply:
> Yes, but the decision to use a dedicated type was inspired by this pattern: https://www.shuttle.rs/blog/2022/07/28/patterns-with-rust-types#contract-based-programming-in-rust--sub-typing

That makes sense for the making sure that the keys/values are valid, but I don't see the value to treating them in isolation.

> I don't know what BTreeSet and BTreeMap has to do with this? We are talking about a single KeyValuePair, which doesn't use any of these collection types. The serialization of the type works very similar to a map, because the identifier is part of the inner data instead the item (struct/enum) name serde uses per default.

`BTreeMap<K, V>` already serializes to a map of:

```yaml
k1: v1
k2: v2
# ...
```

`BTreeSet<(K, V)>` serializes natively to `[[k1, v1], [k2, v2], ...]`, so you do this newtype dance to override both to recreate the map-style serialization. Actually, looking through it again it doesn't look like the `impl Serialize for KeyValuePair` is used as all, since `KeyValuePairs` serializes each key/value separately already?

Reply:
> That makes sense for the making sure that the keys/values are valid, but I don't see the value to treating them in isolation.

Okay, that's why I proposed to implement `Deref<Target = (Key, V)>` or having a function like `.into_inner()` which consumes self and returns `(Key, V)`. Do you have any other suggestions?

> Actually, looking through it again it doesn't look like the `impl Serialize for KeyValuePair` is used as all, since `KeyValuePairs` serializes each key/value separately already?

The `Serialize` for a single `KeyValuePair` is only used, if we want to serialize a single k/v pair, like:

```rust
#[derive(Serialize)]
struct Foo {
    bar: KeyValuePair<LabelValue>,
}
```

Sadly we cannot combine the `Serialize` impl for both a single and multiple k/v pairs.

The serde support might not be necessary, as we most likely wont directly use these structs in our CRDs. So we can opt to remove them for now.

Reply:
I removed serde support for now.